Coerce 01

mattdowle opened the pull request:
Closes #996 
(Closes #3915 too; slightly related regarding NA factor level)
Motivation in the news item.
The complex code removed needed work anyway as it was allocating within a truelength-clobber region. Clearly it's not ideal to making a change as big as this at this stage. On the other hand, it clears up a lot of wrinkles, hopefully.
Several potential coerces removed; e.g. when assigning factor to factor it was coercing the source to character which has gone away now. Plus some other similar todo done.Still some new todo's inline but it's passing locally.  Will run it past revdeps to see the impact.
Interested in first thoughts on this approach.

MichaelChirico commented on src/assign.c:
```diff
@@ -819,89 +702,166 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
+        const int *sD = INTEGER(source);
+        for (int i=0; i<slen; ++i) {
+          const int val = sD[i];
+          if (val!=0 && val!=1 && val!=NA_INTEGER) error("Cannot assign %d to a logical column.", val);
```
could report `i` here... I'm not sure it will always be easy to understand what `i` is referring to though, e.g. assigning subsets

mattdowle replied:
yep. good to do in future.

MichaelChirico commented on src/assign.c:
```diff
@@ -819,89 +702,166 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
+        const double *sD = REAL(source);
+        for (int i=0; i<slen; ++i) {
+          const double d = sD[i];
+          if (!ISNAN(d) && d!=0.0 && d!=1.0) error("Cannot assign %f to a logical column.", d);
```
in the earlier branch you did the NA check last, I guess that will be more efficient in general, do that here as well?

mattdowle replied:
what I had in mind was that if d is nan, then !=0.0 returns nan not true.  But good point. Might be worth coming back to.   ISNAN in R is defined to be C isnan so it's a very fast compare.  I doubt we'd be able to get a measurable difference here -- it's all register ops.

MichaelChirico commented on src/assign.c:
```diff
@@ -819,89 +702,166 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
+        pun = true;
+      }
+      break;
+    case REALSXP:
```
Worth adding punning for `CPLXSXP`? should be similar to `REALSXP`. i can have a go if so.

mattdowle replied:
done.

MichaelChirico commented on src/assign.c:
```diff
@@ -911,16 +871,41 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
+          break;
+        case REALSXP: {
+          const double *sourceD = REAL(source);
+          for (int i=0; i<len; ++i) {
```
the `memcpy` branch is more efficient, is that right? should we report something under `verbose` then?

jangorecki replied:
then we can also test that branch is reached exactly when we expect that by testing verbose output.

mattdowle replied:
yep there's a todo comment in the place to add the verbose/strict option.  Can do it in the next release.

MichaelChirico commented on src/assign.c:
```diff
@@ -982,13 +967,27 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
       } else {
-        const SEXP *val = STRING_PTR(source);
-        for (int i=0; i<len; i++) SET_STRING_ELT(target, start+i, val[i]);
+        if (TYPEOF(source)!=STRSXP) error("Internal error. Not coerced earlier.");
```
nocov

MichaelChirico commented on src/assign.c:
```diff
@@ -1003,27 +1002,81 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
+        }
+      } break;
+      default:
+        error("Internal error");
```
nocov

MichaelChirico commented on src/assign.c:
```diff
@@ -1003,27 +1002,81 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
+        }
+      } break;
+      default:
+        error("Internal error");
```
nocov

MichaelChirico commented on NEWS.md:
```diff
@@ -224,6 +224,8 @@
     # [1] "A" "b" "c"
```
later read from C what's happening, confirmed by Q

mattdowle replied:
I don't understand this one. If something still needs to be done, please follow up as new comment.

jangorecki replied:
fixed in a415095ca382a1b5520027739541ef19ef0dd1a3

mattdowle commented:
Thanks for the great reviews and comments. Seems like a cautious and tentative thumbs up then, depending on revdep results and addressing comments.
5 of 740 packages affected under this PR: `bigreadr` `hutils` `restatapi` `VIM` `wyz.code.testthat`
At least one seems correct: "Cannot assign 0.571429 to a logical column.".  Working through them...

mattdowle commented on src/assign.c:
```diff
@@ -819,89 +702,166 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
+      bool ok = true;
+      switch(TYPEOF(source)) {
+      case LGLSXP: case INTSXP: {
+        const int *sd = INTEGER(source);
```
you mean the `allNA` I wrote in this PR?

MichaelChirico replied:
yes, seems allNA was added but not used?

mattdowle replied:
Yes PR was WIP then; now it uses. I've often wondered why `allNA` didn't seem to exist. So I thought you might have meant we did already have `all_na` or something already (like the request to export `any_na`), or that base/utils already had `all.na` or something somewhere that I didn't know about.

mattdowle commented on inst/tests/tests.Rraw:
```diff
@@ -1218,16 +1216,14 @@ test(423, truelength(DT), 1028L)
-test(426, DT, data.table(a=factor(c("A","Z")),b=1:4))
-DT[1,1] <- 2L
+test(425, DT[1,1] <- 1, error="Cannot assign 'double' to 'factor'")
+test(426, DT[1,1] <- 2L, error="Cannot assign 'integer' to 'factor'")
```
turns out that no revdep fails because of it.  But yes, will reinstate anyway.

mattdowle commented:
Now passes the 2 revdeps that were affected (hutils and VIM). Will do another full rerun after getting full coverage.
Condensed logic using macro.  This gets the salient coerce parts next to each other for easier study.  Overall it's now less lines: diff was +368 - 276 yesterday and now with the macro is +366 - 398.

jangorecki commented on src/assign.c:
```diff
@@ -811,322 +797,201 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source,
+      if (targetIsI64 && isReal(source) && !sourceIsI64) {
+        int firstReal=0;
+        if ((firstReal=INTEGER(isReallyReal(source))[0])) {
+          sprintf(memrecycle_message, "coerced to integer64 but contains a non-integer value (%f at position %d); precision lost.", REAL(source)[firstReal-1], firstReal);
```
shouldn't we use `snprintf` here?

mattdowle replied:
good catch on `snprintf`. done.

> ideally, for 1.13.0 we could try to take out coercion from assign and make it re-usable

that's what I did in this PR iiuc; i.e.  the assign code is taken out of assign() and placed into memrecycle().  memrecycle() is callable from elsewhere.

jangorecki commented on src/assign.c:
```diff
@@ -811,322 +797,201 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source,
+    }                                          \
+  }}
+
+#define COERCE_ERROR(targetType) error("type '%s' cannot be coerced to '%s'", type2char(TYPEOF(source)), targetType); // 'targetType' for integer64 vs double
```
shouldn't be `source` also an argument to `COERCE_ERROR`, it iss being used inside

jangorecki replied:
OK, I understand it is taken from parent scope in `switch` and I assume it never changes

mattdowle replied:
yep this is a little inelegant.  it was because the type of REALSXP doesn't distinguish between "double" and "integer64", which is what I was trying to say by the comment alongside.  We probably do have a slightly higher level function superType() or something which does this.  If not we should create it and replace all type2char(TYPEOF(...)) with superType() instead.

jangorecki commented on src/assign.c:
```diff
@@ -811,322 +797,201 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source,
-      error("Unsupported type in assign.c:memrecycle '%s' (no where)", type2char(TYPEOF(target)));  // # nocov
+
+#undef BODY
+#define BODY(STYPE, RFUN, CTYPE, CAST, ASSIGN) \
```
this is really cool!

jangorecki commented on R/bmerge.R:
```diff
@@ -75,12 +75,12 @@ bmerge = function(i, x, icols, xcols, roll, rollends, nomatch, mult, ops, verbos
         xclass=="logical" || iclass=="logical" ||
         xclass=="factor" || iclass=="factor") {
-      if (anyNA(i[[ic]]) && all(is.na(i[[ic]]))) { # TODO: allNA function in C
+      if (anyNA(i[[ic]]) && allNA(i[[ic]])) {
```
this change is nice but I would hold that for now, we support only few classes in `allNA`, are we sure none of the others can reach here.

mattdowle replied:
good idea.  would agree but I'm not sure what we'd wait for: all tests and all revdeps pass.  So might as well make progress.  Plus is quite a simple function:  hard to get wrong (touch wood!)

jangorecki replied:
@mattdowle I pushed that in extra commit now, could you please re-run revdeps? it was previously returning `false` by default, so the potential issues could have been missed. Now it is more explicit and will raise error.

mattdowle commented:
- [ ] full coverage
- [ ] visit all comment
- [ ] the memcpy todo

mattdowle commented on src/assign.c:
```diff
@@ -489,138 +488,22 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
-    memrecycle(targetcol, rows, 0, targetlen, RHS);  // also called from dogroups where these arguments are used more
-    UNPROTECT(thisprotecti); // unprotect inside loop through columns to save protection stack
+    memrecycle(targetcol, rows, 0, targetlen, thisvalue, coln+1, CHAR(STRING_ELT(names, coln)));
+    // memrecycle is also called from dogroups;          ^^ last 2 arguments just for messages
```
thanks. I removed that part of the comment now.

mattdowle commented on NEWS.md:
```diff
@@ -193,16 +193,16 @@
     x = rnorm(1e6); n = 1e3
     rollfun = function(x, n, FUN) {
-      ans = rep(NA_real_, nx<-length(x))
+      nx = nx = length(x)
```
oops.  I just fixed the extra `nx=`.  My motivaton on this was that my editor was displaying several pages of highlighted green after this section. Something to do with the R syntax highlighting.  Could you make the other changes to that news item as you see fit please.
Why do 2 rows in the timings table have NA in the median column?  Perhaps explain that alongside as a comment to those rows.

jangorecki replied:
I pushed proposed changes.
@mattdowle please re-pull before committing

mattdowle commented on src/assign.c:
```diff
@@ -819,89 +702,166 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
+          if (tl>0) {
+            savetl(s);
+          } else if (tl<0) {
+            // # nocov start
```
Yes.  Just internal errors.  If somehow factor levels contain duplicates (an error because levels must be unique in R),  the 2nd dup would already have our negative truelength assigned to it from the first one (so that's one way to get tl<0). If there are no duplicates then what I had in mind was if core R ever changes their own truelength clobber to be negative (we assume and check that R's internal usage is always positive).   This is what I was trying to get at in the error message: "internal error: levels of target are either not unique or have truelength<0".

mattdowle commented:
Good: the formatting for coverage is doing its job:
![image](https://user-images.githubusercontent.com/2779998/65732513-23f15e00-e07f-11e9-9f06-20a46565cd16.png)

jangorecki commented on src/utils.c:
```diff
@@ -36,29 +36,50 @@ bool allNA(SEXP x) {
+    return false;
+  }
+  default:
+    error("Internal error: unsupported type '%s' passed to allNA()", type2char(TYPEOF(x))); // # nocov
```
We have `raw` and `list` in tests, instead of returning `false` by default we return `false` only for those two, and raise error otherwise. I still see a potential for issues here, for example:
If user defines own class using a list type. Then user can have `is.na.my_class` method, which can work on list, while our `allNA` will not work returning FALSE always.
```
my_cls = structure(list(a=1), class="my_class")
is.na.my_class = function(x) TRUE
```

mattdowle replied:
Good points. `raw` doesn't support NA so that's why `raw`, other than `allNA(raw())` is true and I included test for.  All empty vectors return true.
I wasn't planning on exporting `allNA` in this release.  Just using it internally to start with.
`list`: yes there's some wrinkles there.  `allNA(list())` is correctly true but I don't think I covered a test for that.  But `allNA(list(NA))` is false whereas `all(is.na(list(NA)))` is true in base R. So that needs fixing.  I don't think we pass `list` to `allNA` internally in this PR,  but it should be error at least until it supports it,  otherwise we're just relying on test coverage too much.  Will do.

mattdowle replied:
base R behaviour:
```R
> all(is.na(list(NA, NA)))
[1] TRUE
> all(is.na(list(NA, c(NA))))
[1] TRUE
> all(is.na(list(NA, c(NA,NA))))
[1] FALSE
```

MichaelChirico replied:
there's an explicit

```
	if (!isVector(s) || length(s) != 1)
	    pa[i] = 0;
```

in `do_isna` in `src/main/coerce.c`. No comment as to why... not changed in 19 years

https://github.com/wch/r-source/commit/74bb7a0bdc5e4b21464ec87343f35bf834a6620c

mattdowle replied:
Interesting.  Actually, turning off list support in allNA (so that it errors with unsupported) does cause 3 tests to fail:  1294.05, 1294.12 and 1294.19.   Just in this PR because this PR is now using allNA in bmerge.

mattdowle replied:
And those 3 tests all warn with `Coerced list RHS to logical|integer|double` with on RHS `list(list(1))`.  Comforting that the circle leads back to this `coerce_01` PR.

mattdowle replied:
Since those 3 were warnings already, and with #3626 in mind, I turned them to error and moved on.  It's strange we were allowing these `list` singletons to be coerced to atomic, albeit with warning.  I can see it causing problems with #3626 so those 3 being error make sense for now I hope.

jangorecki commented on src/utils.c:
```diff
@@ -36,29 +36,50 @@ bool allNA(SEXP x) {
   case LGLSXP:
   case INTSXP: {
     const int *xd = INTEGER(x);
-    for (int i=0; i<n; ++i)   if (xd[i]!=NA_INTEGER)   return false;
```
@mattdowle sorry for removing your code alignment. I made it more friendly for codecov now.

mattdowle commented on NEWS.md:
```diff
@@ -192,24 +192,24 @@
+    # zoo::roll[fun]   0.215   0.185      NA   ## median not fully supported
     # frollapply       5.404   1.419  56.475
-    # froll[fun]       0.003   0.002      NA
+    # froll[fun]       0.003   0.002      NA   ## median not yet supported
```
@jangorecki Ahah: I had not guessed that was the reason for the NA (i.e. median not supported).  When I was reading quickly like most people,  I thought it was mean, sum and median of a set of timings,  a bit like microbenchmark output.  Now it makes sense!

mattdowle commented on NEWS.md:
```diff
@@ -221,7 +221,7 @@
     setnames(DT, 2:3, tolower)
     names(DT)
-    # [1] "A" "b" "c"
+    # [1] "a" "b" "c"
```
@jangorecki Why this change?   The `setnames` call sets `2:3` so `c("A","b","c")` was correct.

mattdowle replied:
I guess the 2:3 wasn't clear enough when reading quickly.  I changed it to c(1,3) to make it clearer in 7cad29b

jangorecki replied:
Oh, yes, I haven't notice 2:3! Thanks

MichaelChirico commented on R/utils.R:
```diff
@@ -12,6 +12,7 @@ if (base::getRversion() < "3.5.0") {
 }
 isTRUEorNA    = function(x) is.logical(x) && length(x)==1L && (is.na(x) || x)
 isTRUEorFALSE = function(x) is.logical(x) && length(x)==1L && !is.na(x)
+allNA = function(x) .Call(C_allNAR, x)
```
just ran a quick benchmark and we blow `all(is.na())` out of the water (as expected when `FALSE` but we're also faster when `TRUE` and for moderate vector size).

Worth patching to `base`? It seems not as popular to use as `anyNA` in `base` internal code:

```
grep -Er "all\s*\(is\.na\s*\(" ~/github/r-source --exclude=*tests* | wc -l
# 28, 63 if we include tests code
grep -Er "anyNA\s*\(" ~/github/r-source --exclude=*tests* | wc -l
# 190, 197 if we include tests code
```

MichaelChirico replied:
(as a TODO after release of course)

mattdowle commented on src/utils.c:
```diff
@@ -33,7 +33,7 @@ SEXP isReallyReal(SEXP x) {
 
 bool allNA(SEXP x, bool errorForBadType) {
-  // less space and time than any(is.na(x)) at R level because that creates full size is.na(x) first before any()
+  // less space and time than all(is.na(x)) at R level because that creates full size is.na(x) first before all()
```
oops. thanks @MichaelChirico good spot

mattdowle commented:
More review needed please. Merging to run it through stronger tests in GLCI and so I can move on to the remaining issues in this milestone. Can still add comments in the closed PR and/or open follow-up PRs. Will do another full revdep rerun too.

jangorecki commented:
As of now I see multiple places that could be extended with newlines for C codecov.

jangorecki commented:
we have some strange error not related to this PR
```
ERROR: Preparation failed: Error response from daemon: received unexpected HTTP status: 500 Internal Server Error (executor_docker.go:188:0s)
```
only for R-devel linux image

jangorecki commented:
Seems to be working fine now.
